active/0000-trait-based-exception-handling.md

```diff
@@ -0,0 +1,747 @@
+   automatic exception propagation raises the specter of "exception safety": how
+   serious an issue this would actually be in practice, I don't know - there's
+   reason to believe that it would be much less of one than in C++.
+
+ * Wait (and hope) for HKTs and generic monad sugar.
+
+[notes]: https://github.com/glaebhoerl/rust-notes/blob/268266e8fbbbfd91098d3bea784098e918b42322/my_rfcs/Exceptions.txt
+
+
+# Future possibilities
+
+## Expose a generalized form of `break` or `return` as described
+
+This RFC doesn't propose doing so at this time, but as it would be an independently useful feature, it could be added as well.
+
+## An additional `catch` form to bind the caught exception irrefutably
+
+The `catch` described above immediately passes the caught exception into a
+`match` block. It may sometimes be desirable to instead bind it directly to a
+single variable. That might look like this:
+
+    try { EXPR } catch IRR-PAT { EXPR }
+
+Where `catch` is followed by any irrefutable pattern (as with `let`).
+
+For example:
+
+    try {
+        foo()?.bar()?
+    } catch e {
+        let x = baz(e);
+        quux(x, e);
+    }
+
+While it may appear to be extravagant to provide both forms, there is reason to
+do so: either form on its own leads to unavoidable rightwards drift under some
+circumstances.
+
+The first form leads to rightwards drift if one wishes to do more complex
+multi-statement work with the caught exception:
+
+    try {
+        foo()?.bar()?
+    } catch {
+        e => {
+            let x = baz(e);
+            quux(x, e);
+        }
+    }
+
+This single case arm is quite redundant and unfortunate.
+
+The second form leads to rightwards drift if one wishes to `match` on the caught
+exception:
+
+    try {
+        foo()?.bar()?
+    } catch e {
+        match e {
+            Red(rex)  => baz(rex),
+            Blue(bex) => quux(bex)
+        }
+    }
+
+This `match e` is quite redundant and unfortunate.
+
+Therefore, neither form can be considered strictly superior to the other, and it
+may be preferable to simply provide both.
+
+## `throw` and `throws`
+
+It is possible to carry the exception handling analogy further and also add
+`throw` and `throws` constructs.
+
+`throw` is very simple: `throw EXPR` is essentially the same thing as
+`Err(EXPR)?`; in other words it throws the exception `EXPR` to the innermost
+`try` block, or to the function's caller if there is none.
+
+A `throws` clause on a function:
+
+    fn foo(arg: Foo) -> Bar throws Baz { ... }
+
+would mean that instead of writing `return Ok(foo)` and `return Err(bar)` in the
+body of the function, one would write `return foo` and `throw bar`, and these
+are implicitly turned into `Ok` or `Err` for the caller. This removes syntactic
+overhead from both "normal" and "throwing" code paths and (apart from `?` to
+propagate exceptions) matches what code might look like in a language with
+native exceptions.
+
```

Comment:
I can sympathize with this, and `?` was defined in terms of `throw` in the previous revision. But there's also something to be said for not adding things before we're sure we want them. `Err(foo)?` is weird, but `return Err(foo)` is fine enough; you'd only need the former if you wanted to do it inside a `try` block... how common would that be? Well, we can find out. :)